File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/main/TrackAdapter.java
package mx.dev.franco.automusictagfixer.ui.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.List;

import mx.dev.franco.automusictagfixer.R;
import mx.dev.franco.automusictagfixer.covermanager.CoverLoader;
import mx.dev.franco.automusictagfixer.persistence.room.Track;
import mx.dev.franco.automusictagfixer.utilities.ServiceUtils;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;

public class TrackAdapter extends RecyclerView.Adapter<AudioItemHolder> {

    private static final String TAG = TrackAdapter.class.getName();
    private ServiceUtils serviceUtils;
    private AsyncListDiffer<Track> asyncListDiffer = new AsyncListDiffer<>(this, new DiffCallback());
    private AudioItemHolder.ClickListener mListener;


    public TrackAdapter(){
    }
    public TrackAdapter(AudioItemHolder.ClickListener listener){
        this();
        mListener = listener;
        if (mListener instanceof AsyncListDiffer.ListListener)
            asyncListDiffer.addListListener((AsyncListDiffer.ListListener<Track>) mListener);
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        serviceUtils = ServiceUtils.getInstance(recyclerView.getContext());
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        serviceUtils = null;
        if (mListener instanceof AsyncListDiffer.ListListener)
            asyncListDiffer.removeListListener((AsyncListDiffer.ListListener<Track>) mListener);
        mListener = null;
        CoverLoader.cancelAll();
    }

    /**
     * @inheritDoc
     */
    @Override
    public AudioItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).
                inflate(R.layout.main_item_card, parent, false);
        return new AudioItemHolder(itemView, mListener);
    }

    /**
     * @inheritDoc
     */
    @Override
    public void onBindViewHolder(@NonNull final AudioItemHolder holder, int position,
                                 @NonNull List<Object> payloads) {
        if(payloads.isEmpty()) {
            super.onBindViewHolder(holder,position,payloads);
        }
        else {
            //Track track = mTrackList.get(position);
            Bundle o = (Bundle) payloads.get(0);
                if (o.getString("title") != null)
                    holder.trackName.setText(o.getString("title"));
                if (o.getString("artist") != null)
                    holder.artistName.setText(o.getString("artist"));

                holder.checkBox.setChecked(o.getInt("checked") == 1);


                if (o.getInt("processing") == 1) {
                    holder.progressBar.setVisibility(View.VISIBLE);
                    holder.checkBox.setVisibility(INVISIBLE);
                } else {
                    holder.checkBox.setVisibility(VISIBLE);
                    holder.progressBar.setVisibility(INVISIBLE);
                }

                if (o.getBoolean("should_reload_cover")){
                    loadCover(holder, o.getString("path"), asyncListDiffer.getCurrentList().get(position).getMediaStoreId()+"");
                }
        }
    }

    /**
     * @inheritDoc
     */
    @Override
    public void onBindViewHolder(@NonNull final AudioItemHolder holder, final int position) {
        Track track = asyncListDiffer.getCurrentList().get(position);
        holder.checkBox.setChecked(track.checked() == 1);
        holder.trackName.setText(track.getTitle());
        holder.artistName.setText(track.getArtist());
        if (track.processing() == 1) {
            holder.progressBar.setVisibility(View.VISIBLE);
            holder.checkBox.setVisibility(INVISIBLE);
        } else {
            holder.checkBox.setVisibility(VISIBLE);
            holder.progressBar.setVisibility(INVISIBLE);
        }
        loadCover(holder, track.getPath(), track.getMediaStoreId()+"");
    }

    private void loadCover(AudioItemHolder holder, String path, String mediaStoreId) {
        CoverLoader.startFetchingCover(holder, path, mediaStoreId);
    }

    /**
     * @inheritDoc
     */
    @Override
    public void onViewRecycled(AudioItemHolder holder) {
        if(holder.itemView.getContext() != null)
            Glide.with(holder.itemView.getContext()).clear(holder.cover);
    }

    /**
     * @inheritDoc
     */
    @Override
    public boolean onFailedToRecycleView(@NonNull AudioItemHolder holder) {
        if(holder.itemView.getContext() != null)
            Glide.with(holder.itemView.getContext()).clear(holder.cover);
        return true;
    }

    /**
     * @inheritDoc
     */
    @Override
    public int getItemCount() {
        if(asyncListDiffer != null)
            return asyncListDiffer.getCurrentList().size();
        return 0;
    }

    /**
     * @inheritDoc
     */
    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(true);
    }

    void onChanged(@Nullable List<Track> tracks) {
        asyncListDiffer.submitList(tracks);
    }
}

File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/main/TrackAdapter.kt
package mx.dev.franco.automusictagfixer.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.AsyncListDiffer.ListListener
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import mx.dev.franco.automusictagfixer.R
import mx.dev.franco.automusictagfixer.covermanager.CoverLoader
import mx.dev.franco.automusictagfixer.persistence.room.Track
import mx.dev.franco.automusictagfixer.utilities.ServiceUtils

class TrackAdapter() : RecyclerView.Adapter<AudioItemHolder>(), Observer<List<Track>?> {
    private val serviceUtils: ServiceUtils? = null
    private val asyncListDiffer: AsyncListDiffer<Track>? = AsyncListDiffer(this, DiffCallback())
    private var mListener: AudioItemHolder.ClickListener? = null

    constructor(listener: AudioItemHolder.ClickListener?) : this() {
        mListener = listener
        if (mListener is ListListener<*>) asyncListDiffer!!.addListListener((mListener as ListListener<Track>?)!!)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        mListener = null
        CoverLoader.cancelAll()
    }

    /**
     * @inheritDoc
     */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AudioItemHolder {
        val itemView =
            LayoutInflater.from(parent.context).inflate(R.layout.track_item, parent, false)
        return AudioItemHolder(itemView, mListener)
    }

    /**
     * @inheritDoc
     */
    override fun onBindViewHolder(
        holder: AudioItemHolder, position: Int,
        payloads: List<Any>
    ) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads)
        } else {
            //Track track = mTrackList.get(position);
            val o = payloads[0] as Bundle
            if (o.getString("title") != null) holder.trackName.text = o.getString("title")
            if (o.getInt("processing") == 1) {
                holder.progressBar.visibility = View.VISIBLE
            } else {
                holder.progressBar.visibility = View.INVISIBLE
            }
            if (o.getBoolean("should_reload_cover")) {
                loadCover(
                    holder,
                    o.getString("path"),
                    asyncListDiffer!!.currentList[position].mediaStoreId.toString() + ""
                )
            }
        }
    }

    /**
     * @inheritDoc
     */
    override fun onBindViewHolder(holder: AudioItemHolder, position: Int) {
        val track = asyncListDiffer!!.currentList[position]
        holder.trackName.text = track.title
        //holder.artistName.setText(track.getArtist());
        if (track.processing() == 1) {
            holder.progressBar.visibility = View.VISIBLE
        } else {
            holder.progressBar.visibility = View.INVISIBLE
        }
        loadCover(holder, track.path, track.mediaStoreId.toString() + "")
    }

    private fun loadCover(holder: AudioItemHolder, path: String?, mediaStoreId: String) {
        CoverLoader.startFetchingCover(holder, path, mediaStoreId)
    }

    /**
     * @inheritDoc
     */
    override fun onViewRecycled(holder: AudioItemHolder) {
        if (holder.itemView.context != null) Glide.with(holder.itemView.context).clear(holder.cover)
    }

    /**
     * @inheritDoc
     */
    override fun onFailedToRecycleView(holder: AudioItemHolder): Boolean {
        if (holder.itemView.context != null) Glide.with(holder.itemView.context).clear(holder.cover)
        return true
    }

    /**
     * @inheritDoc
     */
    override fun getItemCount(): Int {
        return asyncListDiffer?.currentList?.size ?: 0
    }

    /**
     * @inheritDoc
     */
    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(true)
    }

    override fun onChanged(tracks: List<Track>?) {
        asyncListDiffer!!.submitList(tracks)
    }

    companion object {
        private val TAG = TrackAdapter::class.java.name
    }
}